File: dp/cpp/mvc.cpp
#include <iostream>
#include <vector>
#include <string>
#include <memory>
#include <algorithm>
#include <optional>

// Model
class User {
private:
    int id_;
    std::string name_;
    std::string email_;

public:
    User(int id, std::string name, std::string email)
        : id_(id), name_(std::move(name)), email_(std::move(email)) {}

    int getId() const { return id_; }
    const std::string& getName() const { return name_; }
    const std::string& getEmail() const { return email_; }

    void setName(const std::string& name) { name_ = name; }
    void setEmail(const std::string& email) { email_ = email; }
};

class UserModel {
private:
    std::vector<std::unique_ptr<User>> users_;

public:
    void addUser(int id, const std::string& name, const std::string& email) {
        users_.push_back(std::make_unique<User>(id, name, email));
    }

    std::optional<std::reference_wrapper<const User>> getUser(int id) const {
        auto it = std::find_if(users_.begin(), users_.end(),
            [id](const auto& user) { return user->getId() == id; });

        if (it != users_.end()) {
            return std::cref(**it);
        }
        return std::nullopt;
    }

    const std::vector<std::unique_ptr<User>>& getAllUsers() const {
        return users_;
    }

    bool removeUser(int id) {
        auto it = std::remove_if(users_.begin(), users_.end(),
            [id](const auto& user) { return user->getId() == id; });

        if (it != users_.end()) {
            users_.erase(it, users_.end());
            return true;
        }
        return false;
    }

    size_t getUserCount() const {
        return users_.size();
    }
};

// View
class UserView {
public:
    virtual ~UserView() = default;

    virtual void displayUser(const User& user) const {
        std::cout << "User ID: " << user.getId() << "\n";
        std::cout << "Name: " << user.getName() << "\n";
        std::cout << "Email: " << user.getEmail() << "\n";
        std::cout << "---\n";
    }

    virtual void displayAllUsers(const std::vector<std::unique_ptr<User>>& users) const {
        std::cout << "=== All Users ===\n";
        for (const auto& user : users) {
            displayUser(*user);
        }
    }

    virtual void displayMessage(const std::string& message) const {
        std::cout << "Message: " << message << "\n";
    }

    virtual void displayError(const std::string& error) const {
        std::cout << "Error: " << error << "\n";
    }

    virtual void displayUserNotFound(int id) const {
        std::cout << "User with ID " << id << " not found\n";
    }
};

// Specialized views
class ConsoleUserView : public UserView {
public:
    void displayUser(const User& user) const override {
        std::cout << "┌─────────────────────────\n";
        std::cout << "│ User ID: " << user.getId() << "\n";
        std::cout << "│ Name: " << user.getName() << "\n";
        std::cout << "│ Email: " << user.getEmail() << "\n";
        std::cout << "└─────────────────────────\n";
    }
};

class JsonUserView : public UserView {
public:
    void displayUser(const User& user) const override {
        std::cout << "{\n";
        std::cout << "  \"id\": " << user.getId() << ",\n";
        std::cout << "  \"name\": \"" << user.getName() << "\",\n";
        std::cout << "  \"email\": \"" << user.getEmail() << "\"\n";
        std::cout << "}\n";
    }
};

// Controller
class UserController {
private:
    std::unique_ptr<UserModel> model_;
    std::unique_ptr<UserView> view_;

public:
    UserController(std::unique_ptr<UserModel> model, std::unique_ptr<UserView> view)
        : model_(std::move(model)), view_(std::move(view)) {}

    void addUser(int id, const std::string& name, const std::string& email) {
        if (name.empty() || email.empty()) {
            view_->displayError("Name and email cannot be empty");
            return;
        }

        // Check if user already exists
        if (model_->getUser(id).has_value()) {
            view_->displayError("User with ID " + std::to_string(id) + " already exists");
            return;
        }

        model_->addUser(id, name, email);
        view_->displayMessage("User added successfully");
    }

    void showUser(int id) {
        auto user = model_->getUser(id);
        if (user.has_value()) {
            view_->displayUser(user->get());
        } else {
            view_->displayUserNotFound(id);
        }
    }

    void showAllUsers() {
        const auto& users = model_->getAllUsers();
        if (users.empty()) {
            view_->displayMessage("No users found");
        } else {
            view_->displayAllUsers(users);
        }
    }

    void removeUser(int id) {
        if (model_->removeUser(id)) {
            view_->displayMessage("User removed successfully");
        } else {
            view_->displayUserNotFound(id);
        }
    }

    void showUserCount() {
        view_->displayMessage("Total users: " + std::to_string(model_->getUserCount()));
    }
};

int main() {
    std::cout << "=== MVC Pattern Demo ===\n\n";

    // Create MVC components
    auto model = std::make_unique<UserModel>();
    auto view = std::make_unique<ConsoleUserView>();
    auto controller = std::make_unique<UserController>(std::move(model), std::move(view));

    // Use controller to manage users
    controller->addUser(1, "Alice Johnson", "alice@example.com");
    controller->addUser(2, "Bob Smith", "bob@example.com");
    controller->addUser(3, "Charlie Brown", "charlie@example.com");

    std::cout << "\n";
    controller->showUserCount();

    // Display specific user
    std::cout << "\nShowing user with ID 2:\n";
    controller->showUser(2);

    // Display all users
    std::cout << "\nShowing all users:\n";
    controller->showAllUsers();

    // Try to show non-existent user
    std::cout << "\nTrying to show user with ID 999:\n";
    controller->showUser(999);

    // Test error handling
    std::cout << "\nTesting error handling:\n";
    controller->addUser(4, "", "invalid@example.com"); // Empty name
    controller->addUser(1, "Duplicate", "duplicate@example.com"); // Duplicate ID

    // Remove user
    std::cout << "\nRemoving user with ID 2:\n";
    controller->removeUser(2);
    controller->showUserCount();

    // Demonstrate different view
    std::cout << "\n=== Using JSON View ===\n";
    auto jsonModel = std::make_unique<UserModel>();
    auto jsonView = std::make_unique<JsonUserView>();
    auto jsonController = std::make_unique<UserController>(std::move(jsonModel), std::move(jsonView));

    jsonController->addUser(1, "John Doe", "john@example.com");
    jsonController->showUser(1);

    return 0;
}

File: dp/rust/src/mvc.rs
/// MVC (Model-View-Controller) Pattern Implementation in Rust
///
/// The MVC pattern separates concerns into three interconnected components:
/// - Model: Manages data and business logic
/// - View: Handles the user interface and presentation
/// - Controller: Acts as an intermediary between Model and View

use std::collections::HashMap;

// Model
#[derive(Debug, Clone)]
pub struct User {
    id: u32,
    name: String,
    email: String,
}

impl User {
    pub fn new(id: u32, name: String, email: String) -> Result<Self, String> {
        if name.is_empty() {
            return Err("Name cannot be empty".to_string());
        }
        if email.is_empty() || !email.contains('@') {
            return Err("Invalid email address".to_string());
        }

        Ok(User { id, name, email })
    }

    pub fn id(&self) -> u32 {
        self.id
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn email(&self) -> &str {
        &self.email
    }

    pub fn set_name(&mut self, name: String) -> Result<(), String> {
        if name.is_empty() {
            return Err("Name cannot be empty".to_string());
        }
        self.name = name;
        Ok(())
    }

    pub fn set_email(&mut self, email: String) -> Result<(), String> {
        if email.is_empty() || !email.contains('@') {
            return Err("Invalid email address".to_string());
        }
        self.email = email;
        Ok(())
    }
}

#[derive(Debug, Default)]
pub struct UserModel {
    users: HashMap<u32, User>,
}

impl UserModel {
    pub fn new() -> Self {
        Self {
            users: HashMap::new(),
        }
    }

    pub fn add_user(&mut self, user: User) -> Result<(), String> {
        if self.users.contains_key(&user.id) {
            return Err(format!("User with ID {} already exists", user.id));
        }
        self.users.insert(user.id, user);
        Ok(())
    }

    pub fn get_user(&self, id: u32) -> Option<&User> {
        self.users.get(&id)
    }

    pub fn get_user_mut(&mut self, id: u32) -> Option<&mut User> {
        self.users.get_mut(&id)
    }

    pub fn remove_user(&mut self, id: u32) -> Option<User> {
        self.users.remove(&id)
    }

    pub fn get_all_users(&self) -> Vec<&User> {
        self.users.values().collect()
    }

    pub fn user_count(&self) -> usize {
        self.users.len()
    }

    pub fn find_users_by_name(&self, name: &str) -> Vec<&User> {
        self.users
            .values()
            .filter(|user| user.name.to_lowercase().contains(&name.to_lowercase()))
            .collect()
    }
}

// View trait for different presentation styles
pub trait UserView {
    fn display_user(&self, user: &User);
    fn display_users(&self, users: &[&User]);
    fn display_message(&self, message: &str);
    fn display_error(&self, error: &str);
    fn display_user_not_found(&self, id: u32);
}

// Console View implementation
pub struct ConsoleUserView;

impl UserView for ConsoleUserView {
    fn display_user(&self, user: &User) {
        println!("┌─────────────────────────");
        println!("│ User ID: {}", user.id());
        println!("│ Name: {}", user.name());
        println!("│ Email: {}", user.email());
        println!("└─────────────────────────");
    }

    fn display_users(&self, users: &[&User]) {
        if users.is_empty() {
            self.display_message("No users found");
            return;
        }

        println!("=== All Users ({}) ===", users.len());
        for user in users {
            self.display_user(user);
        }
    }

    fn display_message(&self, message: &str) {
        println!("✓ {}", message);
    }

    fn display_error(&self, error: &str) {
        println!("✗ Error: {}", error);
    }

    fn display_user_not_found(&self, id: u32) {
        println!("✗ User with ID {} not found", id);
    }
}

// JSON View implementation
pub struct JsonUserView;

impl UserView for JsonUserView {
    fn display_user(&self, user: &User) {
        println!("{{");
        println!("  \"id\": {},", user.id());
        println!("  \"name\": \"{}\",", user.name());
        println!("  \"email\": \"{}\"", user.email());
        println!("}}");
    }

    fn display_users(&self, users: &[&User]) {
        println!("[");
        for (i, user) in users.iter().enumerate() {
            print!("  ");
            if i > 0 {
                println!(",");
                print!("  ");
            }
            println!("{{");
            println!("    \"id\": {},", user.id());
            println!("    \"name\": \"{}\",", user.name());
            println!("    \"email\": \"{}\"", user.email());
            print!("  }}");
        }
        if !users.is_empty() {
            println!();
        }
        println!("]");
    }

    fn display_message(&self, message: &str) {
        println!("{{ \"status\": \"success\", \"message\": \"{}\" }}", message);
    }

    fn display_error(&self, error: &str) {
        println!("{{ \"status\": \"error\", \"message\": \"{}\" }}", error);
    }

    fn display_user_not_found(&self, id: u32) {
        println!("{{ \"status\": \"error\", \"message\": \"User with ID {} not found\" }}", id);
    }
}

// Controller
pub struct UserController<V: UserView> {
    model: UserModel,
    view: V,
}

impl<V: UserView> UserController<V> {
    pub fn new(model: UserModel, view: V) -> Self {
        Self { model, view }
    }

    pub fn add_user(&mut self, id: u32, name: String, email: String) {
        match User::new(id, name, email) {
            Ok(user) => {
                match self.model.add_user(user) {
                    Ok(()) => self.view.display_message("User added successfully"),
                    Err(e) => self.view.display_error(&e),
                }
            }
            Err(e) => self.view.display_error(&e),
        }
    }

    pub fn show_user(&self, id: u32) {
        match self.model.get_user(id) {
            Some(user) => self.view.display_user(user),
            None => self.view.display_user_not_found(id),
        }
    }

    pub fn show_all_users(&self) {
        let users = self.model.get_all_users();
        self.view.display_users(&users);
    }

    pub fn remove_user(&mut self, id: u32) {
        match self.model.remove_user(id) {
            Some(_) => self.view.display_message("User removed successfully"),
            None => self.view.display_user_not_found(id),
        }
    }

    pub fn update_user_name(&mut self, id: u32, new_name: String) {
        match self.model.get_user_mut(id) {
            Some(user) => {
                match user.set_name(new_name) {
                    Ok(()) => self.view.display_message("User name updated successfully"),
                    Err(e) => self.view.display_error(&e),
                }
            }
            None => self.view.display_user_not_found(id),
        }
    }

    pub fn update_user_email(&mut self, id: u32, new_email: String) {
        match self.model.get_user_mut(id) {
            Some(user) => {
                match user.set_email(new_email) {
                    Ok(()) => self.view.display_message("User email updated successfully"),
                    Err(e) => self.view.display_error(&e),
                }
            }
            None => self.view.display_user_not_found(id),
        }
    }

    pub fn search_users(&self, name_query: &str) {
        let users = self.model.find_users_by_name(name_query);
        if users.is_empty() {
            self.view.display_message(&format!("No users found matching '{}'", name_query));
        } else {
            println!("Search results for '{}':", name_query);
            self.view.display_users(&users);
        }
    }

    pub fn show_user_count(&self) {
        self.view.display_message(&format!("Total users: {}", self.model.user_count()));
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_user_creation() {
        let user = User::new(1, "John Doe".to_string(), "john@example.com".to_string()).unwrap();
        assert_eq!(user.id(), 1);
        assert_eq!(user.name(), "John Doe");
        assert_eq!(user.email(), "john@example.com");
    }

    #[test]
    fn test_invalid_user_creation() {
        assert!(User::new(1, "".to_string(), "john@example.com".to_string()).is_err());
        assert!(User::new(1, "John".to_string(), "invalid-email".to_string()).is_err());
    }

    #[test]
    fn test_user_model() {
        let mut model = UserModel::new();
        let user = User::new(1, "John".to_string(), "john@example.com".to_string()).unwrap();

        assert!(model.add_user(user).is_ok());
        assert_eq!(model.user_count(), 1);
        assert!(model.get_user(1).is_some());
        assert!(model.get_user(2).is_none());
    }

    #[test]
    fn test_mvc_integration() {
        let model = UserModel::new();
        let view = ConsoleUserView;
        let mut controller = UserController::new(model, view);

        controller.add_user(1, "Alice".to_string(), "alice@example.com".to_string());
        controller.show_user_count();
    }
}

fn main() {
    println!("=== MVC Pattern Demo ===\n");

    // Create MVC components with Console View
    let model = UserModel::new();
    let view = ConsoleUserView;
    let mut controller = UserController::new(model, view);

    // Add users
    controller.add_user(1, "Alice Johnson".to_string(), "alice@example.com".to_string());
    controller.add_user(2, "Bob Smith".to_string(), "bob@example.com".to_string());
    controller.add_user(3, "Charlie Brown".to_string(), "charlie@example.com".to_string());

    println!();
    controller.show_user_count();

    // Show specific user
    println!("\nShowing user with ID 2:");
    controller.show_user(2);

    // Show all users
    println!("\nShowing all users:");
    controller.show_all_users();

    // Test search functionality
    println!("\nSearching for users with 'Alice':");
    controller.search_users("Alice");

    // Test updates
    println!("\nUpdating user 1's name:");
    controller.update_user_name(1, "Alice Cooper".to_string());
    controller.show_user(1);

    // Test error handling
    println!("\nTesting error handling:");
    controller.add_user(4, "".to_string(), "invalid@example.com".to_string()); // Empty name
    controller.add_user(1, "Duplicate".to_string(), "duplicate@example.com".to_string()); // Duplicate ID

    // Remove user
    println!("\nRemoving user with ID 2:");
    controller.remove_user(2);
    controller.show_user_count();

    // Demonstrate JSON View
    println!("\n=== Using JSON View ===");
    let json_model = UserModel::new();
    let json_view = JsonUserView;
    let mut json_controller = UserController::new(json_model, json_view);

    json_controller.add_user(1, "John Doe".to_string(), "john@example.com".to_string());
    json_controller.show_user(1);
}